src/search/plugins/plugin.h

```diff
@@ -1,103 +1,254 @@
+        const std::string &default_value = "",
+        const Bounds &bounds = Bounds::unlimited(),
+        bool lazy_construction = false);
+
+    template<typename T>
+    void add_list_option(
+        const std::string &key,
+        const std::string &help = "",
+        const std::string &default_value = "",
+        bool lazy_construction = false);
+
+    void document_subcategory(const std::string &subcategory);
+    void document_title(const std::string &title);
+    void document_synopsis(const std::string &note);
+    void document_property(
+        const std::string &property, const std::string &note);
+    void document_language_support(
+        const std::string &feature, const std::string &note);
+    void document_note(
+        const std::string &title, const std::string &note, bool long_text = false);
+
+    const Type &get_type() const;
+    std::string get_key() const;
+    std::string get_title() const;
+    std::string get_synopsis() const;
+    std::string get_subcategory() const;
+    const std::vector<ArgumentInfo> &get_arguments() const;
+    const std::vector<PropertyInfo> &get_properties() const;
+    const std::vector<LanguageSupportInfo> &get_language_support() const;
+    const std::vector<NoteInfo> &get_notes() const;
+};
 
-/*
-  The following function is not meant for users, but only for the
-  plugin implementation. We only declare it here because the template
-  implementations need it.
-*/
-extern void register_plugin_type_plugin(
-    const std::type_info &type,
-    const std::string &type_name,
-    const std::string &documentation);
 
+template<typename Constructed>
+class FeatureWithDefault : public Feature {
+protected:
+    using Feature::Feature;
+    virtual std::shared_ptr<Constructed> create_component(
+        const Options &options, const utils::Context &) const {
+        return std::make_shared<Constructed>(options);
+    }
+};
+
+template<typename Constructed>
+class FeatureWithoutDefault : public Feature {
+protected:
+    using Feature::Feature;
+    virtual std::shared_ptr<Constructed> create_component(
+        const Options &, const utils::Context &) const = 0;
+};
+
+template<typename Constructed>
+using FeatureAuto = typename std::conditional<
+    std::is_constructible<Constructed, const Options &>::value,
+    FeatureWithDefault<Constructed>,
+    FeatureWithoutDefault<Constructed>>::type;
+
+template<typename Base, typename Constructed>
+class TypedFeature : public FeatureAuto<Constructed> {
+    using BasePtr = std::shared_ptr<Base>;
+    static_assert(std::is_base_of<Base, Constructed>::value,
+                  "Constructed must derive from Base");
+public:
+    TypedFeature(const std::string &key)
+        : FeatureAuto<Constructed>(TypeRegistry::instance()->get_type<BasePtr>(), key) {
+    }
+
+    Any construct(const Options &options, const utils::Context &context) const override {
+        std::shared_ptr<Base> ptr = this->create_component(options, context);
```

Comment:
This question came up somewhere else later ( https://github.com/aibasel/downward/pull/245#discussion_r1951068507 ) but it actually belongs here:

Q: (Why) do we need `this` here?
A: It is needed
https://isocpp.org/wiki/faq/templates#nondependent-name-lookup-members